lib/Fmt_ast.ml

```diff
@@ -19,12 +19,48 @@ open Extended_ast
 open Ast
 open Fmt
 
+module Layout_cache = struct
+  module Key = struct
+    type t = Arg of arg_label * expression | Pattern of pattern
+
+    let expression_to_string e =
+      Caml.Format.asprintf "%a" Pprintast.expression e
+
+    let pattern_to_string e = Caml.Format.asprintf "%a" Pprintast.pattern e
+
+    let sexp_of_arg_label = function
+      | Nolabel -> Sexp.Atom "Nolabel"
+      | Labelled label -> List [Atom "Labelled"; sexp_of_string label]
+      | Optional label -> List [Atom "Optional"; sexp_of_string label]
+
+    let sexp_of_t = function
+      | Arg (label, expression) ->
+          Sexp.List
+            [ Atom "Arg"
+            ; sexp_of_arg_label label
+            ; sexp_of_string (expression_to_string expression) ]
+      | Pattern pattern ->
+          List [Atom "Pattern"; sexp_of_string (pattern_to_string pattern)]
+
+    let compare = Poly.compare
+
+    let hash = Hashtbl.hash
+  end
+
+  let create () = Hashtbl.create (module Key)
+
+  type t = string Hashtbl.M(Key).t
+
+  let memo t ~key ~f = Hashtbl.find_or_add t key ~default:f
+end
+
 type c =
   { conf: Conf.t
   ; debug: bool
   ; source: Source.t
   ; cmts: Cmts.t
-  ; fmt_code: Conf.t -> string -> (Fmt.t, unit) Result.t }
+  ; fmt_code: Conf.t -> string -> (Fmt.t, unit) Result.t
+  ; layout_cache: Layout_cache.t }
```

Comment:
I'm not sure it's worth to add this field to the type c, as we don't need to pass it to every function, it's only used by Cmts.preserve, why not add it to Cmts.t and directly modify the implementation of Cmts.preserve?

Reply:
I did so because it's not really related to comment handling, but yes we can do that.